src/lib/query-string.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * Encodes string by stricter rules
+ * see https://github.com/kevva/strict-uri-encode
+ *
+ * @param {String} string
+ */
+const encode = string => encodeURIComponent(string).replace(/[!'()*]/g, x => `%${x.charCodeAt(0).toString(16).toUpperCase()}`);
+
+/**
+ * Returns an array, where first element is the slice till separator and the second is the rest.
+ *
+ * @param {String} string
+ * @param {String} separator
+ */
+const splitOnFirst = (string, separator) => {
+    const separatorIndex = string.indexOf(separator);
+    if (separatorIndex === -1) {
+        return [string];
+    }
+    return [
+        string.slice(0, separatorIndex),
+        string.slice(separatorIndex + separator.length),
+    ];
+};
+
+/**
+ * Sorts array or object by keys
+ *
+ * @param {Array|Object} input
+ */
+const keysSorter = (input) => {
+    if (Array.isArray(input)) {
+        return input.sort();
+    }
+
+    if (typeof input === 'object') {
+        return keysSorter(Object.keys(input))
+            .sort((a, b) => Number(a) - Number(b))
+            .map(key => input[key]);
+    }
+
+    return input;
+};
+
+/**
+ * Formats the object by adding new keys
+ *
+ * @param {String} key
+ * @param {String} value
+ * @param {Object} accumulator
+ */
+const decoderFormatter = (key, value, accumulator) => {
+    if (accumulator[key] === undefined) {
+        accumulator[key] = value;
+        return;
+    }
+
+    accumulator[key] = [].concat(accumulator[key], value);
+};
+
+/**
+ * Reducer that formats and combines key-value pair
+ *
+ * @param {String} key
+ */
+const encoderFormatter = key => (result, value) => {
+    if (value === undefined) {
+        return result;
+    }
+
+    if (value === null) {
+        return [...result, encode(key)];
+    }
+
+    return [...result, [encode(key), '=', encode(value)].join('')];
+};
+
+/**
+ * Converts query string into object
+ *
+ * @param {String} input
+ */
+export function parse(input) {
+    const ret = Object.create(null);
+
+    if (typeof input !== 'string') {
+        return ret;
+    }
+
+    const inputParsed = input.trim().replace(/^[?#&]/, '');
+
+    if (!inputParsed) {
+        return ret;
+    }
+
+    inputParsed.split('&').forEach(param => {
+        /* eslint-disable prefer-const */
+        let [key, value] = splitOnFirst(param.replace(/\+/g, ' '), '=');
+
+        value = value === undefined ? null : decodeURIComponent(value);
+        decoderFormatter(decodeURIComponent(key), value, ret);
+    });
+
+    return Object.keys(ret).sort().reduce((result, key) => {
```

Comment:
Just curious: why do we need to sort the keys?

Reply:
query-string has sort by default enabled and we were using. So it's for backward compatibility

Reply:
Ok, got it, thanks